src/mscorlib/shared/System/StringComparer.cs

```diff
@@ -167,14 +176,28 @@ public sealed class CultureAwareComparer : StringComparer
     {
         private readonly CompareInfo _compareInfo; // Do not rename (binary serialization)
         private readonly bool _ignoreCase; // Do not rename (binary serialization)
+        private CompareOptions _compareOptions;
 
```

Comment:
I believe adding this breaks serialization compat with desktop. It likely needs to be made Optional.

Reply:
Good catch, yes it doesn't implement ISerializable. @ViktorHofer I guess this is the first point we have to consider cross version serialization from 2.0. Should we use `[OptionalField]` (with a Version?) and have the code check _ignoreCase first? Or switch this to ISerializable (which I assume could be done non breaking, but I don't know.) so we can set it to match _ignoreCase.

Reply:
Also @ViktorHofer do we have a test strategy for 2.0<->2.1 serialization interop?

Reply:
Another option is one of the [De]Serializ[ing|ed] attributes, but IIRC there was some issue with those that we fixed recently (?)

@Anipik if you're not famliar with binary serialization, read all [these pages](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binary-serialization) including [this one that is relevant here](https://docs.microsoft.com/en-us/dotnet/standard/serialization/version-tolerant-serialization)

Reply:
Correct, the following would happen:

Core 2.1 --> Core2.0/NETFX: pass, additional field in the blob is ignored
NETFX/Core2.0 --> Core 2.1: throw, missing field in the blob

The easiest solution is as Stephen stated mark `_compareOptions` as optional. IMO the best solution would be to mark the type as ISerializable, get rid of the `_ignoreCase` field and during deserialization try to read `_compareOptions` from the serialization blob with `GetValueNoThrow` and if that fails read `_ignoreCase` from the blob. For serialization inside GetObjectData, both fields would needed to be written to the blob.

Reply:
> Also @ViktorHofer do we have a test strategy for 2.0<->2.1 serialization interop?

Not today. We haven't had such a case yet. But is fairly easy to implement. Just add another item to the array in the test data and change 1-2 lines which indicate which index is which platform.

Reply:
I guess it needs a version ie  `[OptionalField(VersionAdded = 2)] ` at least according to [the doc](https://docs.microsoft.com/en-us/dotnet/standard/serialization/version-tolerant-serialization)

Reply:
@Anipik can you add such a test that @ViktorHofer describes? He can point you at the serialization tests and give you a pointer.

Reply:
> I guess it needs a version ie [OptionalField(VersionAdded = 2)] at least according to the doc

The version is not used by the runtime. it is just a hint for the dev. I wouldn't add the version

Reply:
Why not making it ISerializable instead?...

Reply:
please hold on as we are investigating another issue on the full framework. we may end up using ISerializable

Reply:
to clarify, ISerializable is usually the old way used in serialization and it is used more if you are changing the field names. anyway, we need to wait a little bit to tell what is the best way we should go with.

Reply:
one more thing, OptionalField is very important to have, otherwise will break deserialization on old desktop versions (4.7 for instance)

Reply:
We are using it on plenty of types in the BCL, e.g. on all Exception types when we need more control over the serialization process. In this case it makes sense as we can get rid of the extra bool fields which benefits all scenarios, even when you don't care about serialization at all

Reply:
> we can get rid of the extra bool fields which benefits all scenarios

This will break the deserialization on on the full framework. except if you are talking to manually store it when we are serializing it. either way, we have to handle it.

Reply:
Yes, I mean to manually store it, that's why I'm pushing for ISerializable.